src/Adapter/Product/Image/Repository/ProductImageMultiShopRepository.php

```diff
@@ -0,0 +1,242 @@
+            $images[] = $this->get($imageId, $shopId);
+        }
+
+        return $images;
+    }
+
+    /**
+     * @param ProductId $productId
+     *
+     * @return ImageId[]
+     *
+     * @throws CoreException
+     */
+    public function getImagesIds(ProductId $productId, ShopConstraint $shopConstraint): array
+    {
+        $qb = $this->connection->createQueryBuilder()->select('id_image');
+
+        //todo: manage position
+        if ($shopConstraint->getShopGroupId()) {
+            throw new InvalidShopConstraintException('Image has no features related with shop group use single shop and all shops constraints');
+        } elseif ($shopConstraint->forAllShops()) {
+            $qb->from($this->dbPrefix . 'image', 'i')
+                ->addOrderBy('i.position', 'ASC')
+            ;
+        } else {
+            $qb->from($this->dbPrefix . 'image_shop', 'i')
+                ->andWhere('i.id_shop = :shopId')
+                ->setParameter('shopId', $shopConstraint->getShopId()->getValue())
+            ;
+        }
+
+        $results = $qb->andWhere('i.id_product = :productId')
+            ->setParameter('productId', $productId->getValue())
+            ->addOrderBy('i.id_image', 'ASC')
+            ->execute()
+            ->fetchAll()
+        ;
+
+        if (!$results) {
+            return [];
+        }
+
+        $imagesIds = [];
+        foreach ($results as $result) {
+            $imagesIds[] = new ImageId((int) $result['id_image']);
+        }
+
+        return $imagesIds;
+    }
+
+    /**
+     * @param ImageId $imageId
+     *
+     * @return Image
+     *
+     * @throws CoreException
+     */
+    public function get(ImageId $imageId, ShopId $shopId): Image
+    {
+        /** @var Image $image */
+        $image = $this->getObjectModelForShop(
+            $imageId->getValue(),
+            Image::class,
+            ProductImageNotFoundException::class,
+            $shopId
+        );
+
+        return $image;
+    }
+
+    /**
+     * @param ImageId $imageId
+     *
+     * @return ShopId[]
+     */
+    public function getAssociatedShopIds(ImageId $imageId): array
+    {
+        $qb = $this->connection->createQueryBuilder();
+        $qb
+            ->select('id_shop')
+            ->from($this->dbPrefix . 'image_shop')
+            ->where('id_image = :imageId')
+            ->setParameter('imageId', $imageId->getValue())
+        ;
+
+        $result = $qb->execute()->fetchAll();
+        if (empty($result)) {
+            return [];
+        }
+
+        $shops = [];
+        foreach ($result as $shop) {
+            $shops[] = new ShopId((int) $shop['id_shop']);
+        }
+
+        return $shops;
+    }
+
+    public function create(ProductId $productId, ShopConstraint $shopConstraint): Image
+    {
+        $productIdValue = $productId->getValue();
+        $image = new Image();
+        $image->id_product = $productIdValue;
+        $image->cover = !Image::getCover($productIdValue);
```

Comment:
The cover is not to handle in this PR, but I think here there will be a problem. You should check if the product has a cover for the specific shop constraint, for now it tests that the product has a cover for any shop, which means some shop will end up without a cover potentially (as long as one has a cover) So be careful on this one in the next PR

Reply:
should add a todo in code (anyway there are some changes to apply for this PR)